### PR TITLE
Sol compiletime check for out of bound index(integer constant) access for Ordinary arrays

### PR DESCRIPTION
**Pull request by LianaHus**
PR 1 for https://www.pivotaltracker.com/n/projects/1189488/stories/97908844

### CODE REVIEW

File: libsolidity/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -855,8 +857,20 @@ bool ExpressionCompiler::visit(IndexAccess const& _indexAccess)
 		solAssert(_indexAccess.getIndexExpression(), "Index expression expected.");
 
 		_indexAccess.getIndexExpression()->accept(*this);
+
+		// check for dynamically sized arrays should be done after memberAccess visit to have length
+		if (
+			(indexType.getCategory() == Type::Category::IntegerConstant) &&
+				((arrayType.isDynamicallySized() && arrayType.getLength()) || !arrayType.isDynamicallySized()))
```
Comment (gavofyork) [inline libsolidity/ExpressionCompiler.cpp]: one fewer tabs and include final `)` on own line, like opening `(`.

File: libsolidity/ExpressionCompiler.cpp [hunk 2]
```diff
@@ -855,8 +857,21 @@ bool ExpressionCompiler::visit(IndexAccess const& _indexAccess)
 		solAssert(_indexAccess.getIndexExpression(), "Index expression expected.");
 
 		_indexAccess.getIndexExpression()->accept(*this);
+
```
Comment (chriseth) [inline libsolidity/ExpressionCompiler.cpp]: Please do this check as part of the type-checking phase, no the code generation phase. I.e. it probably has to go AST.cpp.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 3]
```diff
@@ -1037,12 +1037,9 @@ BOOST_AUTO_TEST_CASE(array_accessor)
 	)";
 	compileAndRun(sourceCode);
 	BOOST_CHECK(callContractFunction("data(uint256)", 0) == encodeArgs(8));
-	BOOST_CHECK(callContractFunction("data(uint256)", 8) == encodeArgs());
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp:1040]: These should still stay, they are run-time checks.

File: libsolidity/AST.cpp [hunk 4]
```diff
@@ -113,8 +113,7 @@ map<FixedHash<4>, FunctionTypePointer> ContractDefinition::interfaceFunctions()
 	for (auto const& it: exportedFunctionList)
 		exportedFunctions.insert(it);
 
-	solAssert(
-		exportedFunctionList.size() == exportedFunctions.size(),
+	solAssert(exportedFunctionList.size() == exportedFunctions.size(),
```
Comment (chriseth) [inline libsolidity/AST.cpp]: This one was correct before.

File: libsolidity/AST.cpp [hunk 5]
```diff
@@ -1086,10 +1097,12 @@ void IndexAccess::checkTypeRequirements(TypePointers const*)
 		if (type.isString())
 			BOOST_THROW_EXCEPTION(createTypeError("Index access for string is not possible."));
 		m_index->expectType(IntegerType(256));
-		if (type.isByteArray())
-			m_type = make_shared<FixedBytesType>(1);
-		else
-			m_type = type.baseType();
+
+		m_type = type.baseType();
+		if(IntegerConstantType const* integerType = dynamic_cast<IntegerConstantType const*>(m_index->type().get()))
```
Comment (chriseth) [inline libsolidity/AST.cpp]: You can use `auto integerType =` here and please add a space after the `if`.

File: libsolidity/ExpressionCompiler.cpp [hunk 6]
```diff
@@ -856,8 +858,21 @@ bool ExpressionCompiler::visit(IndexAccess const& _indexAccess)
 		solAssert(_indexAccess.indexExpression(), "Index expression expected.");
 
 		_indexAccess.indexExpression()->accept(*this);
+
```
Comment (chriseth) [inline libsolidity/ExpressionCompiler.cpp]: You can remove this here, we already checked during the analysis phase.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 7]
```diff
@@ -4180,22 +4182,6 @@ BOOST_AUTO_TEST_CASE(evm_exceptions_in_constructor_call_fail)
 	BOOST_CHECK(callContractFunction("test()") == encodeArgs(2));
 }
 
-BOOST_AUTO_TEST_CASE(evm_exceptions_in_constructor_out_of_baund)
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp:4183]: We could perhaps change this to a check that the compiler cannot detect at compile-time, e.g. you can assign 5 to a temporary variable.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 8]
```diff
@@ -5230,6 +5233,20 @@ BOOST_AUTO_TEST_CASE(storage_string_as_mapping_key_without_variable)
 	BOOST_CHECK(callContractFunction("f()") == encodeArgs(u256(2)));
 }
 
+BOOST_AUTO_TEST_CASE(array_out_of_bound_access)
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Please move this to SolidityNameAndTypeResolution, as it tests the analysis phase.

**Comment by chriseth** [comment]
Please rebase.

**Pull request merged**